Lib/test/test_iter.py

```diff
@@ -238,6 +247,81 @@ def test_mutating_seq_class_exhausted_iter(self):
         self.assertEqual(list(empit), [5, 6])
         self.assertEqual(list(a), [0, 1, 2, 3, 4, 5, 6])
 
+    def test_reduce_mutating_builtins_iter(self):
+        # This is a reproducer of issue #101765
+        # where iter `__reduce__` calls could lead to a segfault or SystemError
+        # depending on the order of C argument evaluation, which is undefined
+
+        # Backup builtins
+        builtins_dict = builtins.__dict__
+        orig = {"iter": iter, "reversed": reversed}
+
+        def run(builtin_name, item, init=None, sentinel=None):
+            if init is not None:
+                args = init if isinstance(init, tuple) else (init,)
```

Comment:
nit, but this seems like unnecessary magic for a test case; i'm not sure it really increases readability/clarity of the cases vs just letting them all be literal tuples, mostly of length 1, given that it means you have to understand that `()` means something very different from any other single value

Reply:
Yeah, in retrospect I don't think we need to construct the actual object within the function anyways, I've simplified it to just
```python
def run(builtin_name, item, sentinel=None):
    it = iter(item) if sentinel is None else iter(item, sentinel)
    ...
```
in [19ab9c6](https://github.com/python/cpython/pull/101769/commits/19ab9c63b275eaa3b6e10e04d5eec642083f233f)